stan/math/prim/fun/unit_vector_constrain.hpp

```diff
@@ -47,13 +44,12 @@ inline plain_type_t<T> unit_vector_constrain(const T& y) {
 template <typename T1, typename T2, require_eigen_col_vector_t<T1>* = nullptr,
           require_all_not_vt_autodiff<T1, T2>* = nullptr>
 inline plain_type_t<T1> unit_vector_constrain(const T1& y, T2& lp) {
-  using std::sqrt;
   check_nonzero_size("unit_vector_constrain", "y", y);
   auto&& y_ref = to_ref(y);
-  value_type_t<T1> SN = dot_self(y_ref);
-  check_positive_finite("unit_vector_constrain", "norm", SN);
-  lp -= 0.5 * SN;
-  return y_ref.array() / sqrt(SN);
+  value_type_t<T1> r = norm2(y_ref);
+  check_positive_finite("unit_vector_constrain", "norm", r);
+  lp -= 0.5 * (r * r);
```

Comment:
This looks like it's also just paying that squaring cost again here on line 51, so I think the original code may again be faster here.

What we really want here is a specialized function for unit_vector_constrain that does all of this with analytic derivatives for lp and the return type.